Add Code actions on save

junglerobba opened the pull request:
For now this is just https://github.com/helix-editor/helix/pull/6486 redone on top of latest master. Because there was no more discussion on that PR I thought I’d resubmit a working recent version and see if there’s anything left to do.
The only thing changed is making it work akin to autoformat so that code actions don’t trigger on auto save.

junglerobba committed:
lsp: add code actions on save

Adds a new `code-actions-on-save` config option for LSP code actions to
run on save, before auto formatting will be applied.

These actions will be spawned as a series of alternating async jobs and
callbacks, using the callback followups from the previous commit, which
starts the next async task after each callback executes.
This is needed to ensure these actions each operate on the latest
document version and will run in the configured order, so access to
non-thread-safe Editor is required for querying document state in
between async tasks.

Because this is run automatically on save, and is user-configurable, the
reasonable default is using each action from the first LSP that
advertises it and, in case it resolves to multiple actions, use the
first one. This way no user interaction is required, and potentially
conflicting applications of actions are avoided.

Co-authored-by: Jonatan Pettersson <jonatan.pettersson@proton.me>

junglerobba committed:
commands: skip code actions on auto save

Adds a new `--no-code-actions` flag for the write and write-all commands
to disable configured on save code actions.

archseer reviewed (changes requested):
Sorry this has been stuck in the queue for a while but it looks good to me apart from a small nit!

archseer reviewed (dismissed):
Looks good to me now, I'll test drive this for a couple of days :)

junglerobba commented:
Since the largest part of this is still just the original PR rebased, I hope I split the attributions to the original author correctly between commits :)

junglerobba committed:
jobs: allow callbacks to return followup jobs

This introduces `Callback::Followup` that can optionally return
another job to be queued after the callback finishes.
This allows creation of async operation chains where the editor state
needs to be queried in between async operations, without needing
blocking operations.

Needed for features like running multiple LSP code actions in order,
where each action must operate on the latest document state after the
previous action completes.

the-mikedavis commented on helix-term/src/commands/lsp.rs:
```diff
@@ -825,6 +797,155 @@ pub fn code_action(cx: &mut Context) {
+    let mut resolved_code_action = None;
+    if code_action.edit.is_none() || code_action.command.is_none() {
+        if let Some(future) = language_server.resolve_code_action(code_action) {
+            if let Ok(code_action) = helix_lsp::block_on(future) {
```
`format_selections` has a good example of how this is usually done https://github.com/helix-editor/helix/blob/b7fbae574b15ece8094e191e2d797201c76d1296/helix-term/src/commands.rs#L5074-L5094

You can spawn a tokio task to await the future off of the main thread and then use `job::dispatch` to modify the editor with the result on the main thread.

junglerobba commented on helix-lsp/src/client.rs:
```diff
@@ -517,6 +518,27 @@ impl Client {
         }
     }
 
+    pub fn notify_sync<R: lsp::notification::Notification>(&self, params: R::Params) -> Result<()>
```
Dropped this part 👍

the-mikedavis commented on helix-term/src/commands/typed.rs:
```diff
@@ -498,6 +516,7 @@ fn write(cx: &mut compositor::Context, args: Args, event: PromptEvent) -> anyhow
         WriteOptions {
             force: false,
             auto_format: !args.has_flag(WRITE_NO_FORMAT_FLAG.name),
+            code_actions: !args.has_flag(WRITE_NO_FORMAT_FLAG.name),
```
I think we should introduce a new flag for this, `--no-format` skipping code actions seems potentially unexpected to me

the-mikedavis committed:
lsp: Restructure code-actions-on-save as a flat chain

The on-save chain was built by folding the configured kinds in reverse
into nested `Followup` closures, each calling `queued_job.take()`. That
reverse construction is what produced the `wait_futures` bug earlier in
the branch history.

This change replaces it with a flat state machine.
`code_action_on_save_step` carries the remaining kinds
(`VecDeque<CodeActionKind>`) plus the trailing format/save job, popping
one kind per link. Each link is still a lazily built `Followup` so it
reads the document after the previous link's edits and runs the kinds in
configured order.

Resolution now reuses `resolve_code_action_blocking` (as the interactive
picker does), dropping the separate async `resolve_code_action` /
`make_code_action_resolve_callback` layer.

Co-authored-by: Blaž Hrastnik <blaz@mxxn.io>

the-mikedavis committed:
lsp: Apply every code action matching a configured kind on save

Two bugs fixed:

* Only the first returned action was applied. VS Code's
  `codeActionsOnSave` applies every action of each configured kind, and
  it matters for the `source.fixAll` family: eslint/ruff/tsserver expose
  `source.fixAll.eslint`, `source.fixAll.ts`, etc., as separate actions
  under the umbrella, and a user who configures `source.fixAll` wants
  all of them. First-only silently dropped the rest.
* The `only` request field is just a server-side hint; some servers
  ignore it or return a hierarchical superset. We applied whatever came
  back with no client-side check, so a server returning an unrelated
  action had it applied silently, making the config not mean what it
  says.

Apply all returned actions whose kind matches the requested kind under
the LSP kind hierarchy (`source.fixAll.eslint` matches `source.fixAll`),
via the new `code_action_kind_matches`.

Co-authored-by: Blaž Hrastnik <blaz@mxxn.io>

the-mikedavis committed:
lsp: Skip stale code-actions-on-save edits when the document changed

The workspace-edit apply was not version-guarded, unlike the auto-format
path. Across the request->resolve->apply round-trips the buffer can
advance (the user keeps typing during a slow chain, or other jobs
interleave on `:wq`) and applying an edit at positions that no longer
exist corrupts the buffer.

Capture the document version when the kind's request is issued and bail
out of the apply (skipping that kind, continuing the chain) if the active
document's version changed in the meantime.

Co-authored-by: Blaž Hrastnik <blaz@mxxn.io>

the-mikedavis committed:
commands: Always save on write when code actions run w/o auto-format

The on-save chain's tail was the auto-format job, which is `None` when
auto-format is off. With code actions configured but auto-format
disabled (globally, or per write via `:w --no-format`), the chain
applied the actions and then ran an empty tail, so the file was never
written. The actions ran but the save was silently dropped.

Build the tail whenever there is any pre-save work (auto-format or
code actions) and have it always save: run the formatter if there
is one, else write directly. A plain `:w` with neither still saves
synchronously as before.

Co-authored-by: Blaž Hrastnik <blaz@mxxn.io>

the-mikedavis reviewed (approved):
Sorry for the hold-up! I wanted to see if we could eliminate `Followup` but I don't think we can while also still blocking correctly on exit. Eventually we want to move document saving to the event system and maybe then we could drop `Followup`. @archseer and I gave this another pass and pushed a few fixes

the-mikedavis commented:
Thanks!

sulgg commented:
Hi, I'm just testing `code-actions-on-save` on python code, and it looks like is not working. This is my conf:
```toml
[[language]]
name = "python"
language-servers = ["ty", "ruff"]
auto-format = true
code-actions-on-save = [
  "source.organizeImports",
  "source.fixAll",
]
```
I'm not a Rust expert, so maybe if it helps, this is what my code agent says:
"This is a job chain — it builds a Job that runs before the actual write. If the save command (:w) isn't passing through this job chain, the actions never run."

bojohnson5 commented:
> Hi, I'm just testing `code-actions-on-save` on python code, and it looks like is not working. This is my conf:
> 
> ```toml
> [[language]]
> name = "python"
> language-servers = ["ty", "ruff"]
> auto-format = true
> code-actions-on-save = [
>   "source.organizeImports",
>   "source.fixAll",
> ]
> ```
> 
> I'm not a Rust expert, so maybe if it helps, this is what my code agent says: "This is a job chain — it builds a Job that runs before the actual write. If the save command (:w) isn't passing through this job chain, the actions never run."

@sulgg I had the same issue. If you switch the order of your language servers, it _should_ work as you intend, allowing `ruff` to run its code actions. I believe the issue stems from this PR only taking the first LSP that matches a code action and doesn't consider that languages like Python have multiple LSPs with potential code actions. @junglerobba is that a fair way to assess what's happening with the issue above?

sulgg commented:
Thanks @bojohnson5, it works now. I should have read the commit descriptions :(

junglerobba commented:
Not familiar with python tooling, but if you need a specific one to run a code action when multiple are matching it, it should be possible to use `"source.fixAll.ruff"` and `"source.organizeImports.ruff"`.
At least going by [their docs](https://docs.astral.sh/ruff/editors/features/#code-actions)

bojohnson5 commented:
> Not familiar with python tooling, but if you need a specific one to run a code action when multiple are matching it, it should be possible to use `"source.fixAll.ruff"` and `"source.organizeImports.ruff"`. At least going by [their docs](https://docs.astral.sh/ruff/editors/features/#code-actions)

@junglerobba With this in `languages.toml` I just confirmed that the `ruff` code actions aren't run when it is listed 2nd.

```
[[language]]
name = "python"
auto-format = true
language-servers = ["ty", "ruff"]
code-actions-on-save = ["source.organizeImports.ruff"]
```
but it does work, with and without the `.ruff` suffix, when `ruff` comes first in `language-servers`.

I'm totally fine with putting `ruff` first in the list, I just don't know enough about the helix codebase or Rust to understand if this is expected or not.
